jetty-webapp/src/test/java/org/eclipse/jetty/webapp/ForcedServletTest.java

```diff
@@ -0,0 +1,269 @@
+        // status code 555 is from RejectUncompiledJspServlet
+        assertEquals(555, response.getStatus());
+    }
+
+    /**
+     * Test access of a jsp resource that does not exist in the base resource, but in the descriptor, as a precompiled jsp entry
+     */
+    @Test
+    public void testPrecompiledEntry() throws Exception
+    {
+        StringBuilder rawRequest = new StringBuilder();
+        rawRequest.append("GET /precompiled/world.jsp HTTP/1.1\r\n");
+        rawRequest.append("Host: local\r\n");
+        rawRequest.append("Connection: close\r\n");
+        rawRequest.append("\r\n");
+
+        String rawResponse = connector.getResponse(rawRequest.toString());
+        HttpTester.Response response = HttpTester.parseResponse(rawResponse);
+        assertEquals(200, response.getStatus());
+
+        String responseBody = response.getContent();
+        assertThat(responseBody, containsString("This is the FakePrecompiledJSP"));
+    }
+
+    public static class TestInit extends AbstractLifeCycle implements ServletContextHandler.ServletContainerInitializerCaller
+    {
+        private final WebAppContext _webapp;
+
+        public TestInit(ForcedWebAppContext webapp)
+        {
+            _webapp = webapp;
+        }
+
+        @Override
+        protected void doStart() throws Exception
+        {
+            // This will result in a 404 for all requests that don't belong to a more precise servlet
+            forceServlet("default", ServletHandler.Default404Servlet.class);
+            addServletMapping("default", "/");
+
+            // This will result in any attempt to use an JSP that isn't precompiled and in the descriptor with status code 555
+            forceServlet("jsp", RejectUncompiledJspServlet.class);
+            addServletMapping("jsp", "*.jsp");
+            super.doStart();
+        }
+
+        private void addServletMapping(String name, String pathSpec)
+        {
+            ServletMapping mapping = new ServletMapping();
+            mapping.setServletName(name);
+            mapping.setPathSpec(pathSpec);
+            _webapp.getServletHandler().addServletMapping(mapping);
+        }
+
+        private void forceServlet(String name, Class<? extends HttpServlet> servlet) throws Exception
+        {
+            ServletHandler handler = _webapp.getServletHandler();
+
+            // Remove existing holder
+            handler.setServlets(Arrays.stream(handler.getServlets())
+                .filter(h -> !h.getName().equals(name))
+                .toArray(ServletHolder[]::new));
+
+            // add the forced servlet
+            ServletHolder holder = new ServletHolder(servlet.getConstructor().newInstance());
+            holder.setInitOrder(1);
+            holder.setName(name);
+            holder.setAsyncSupported(true);
+            handler.addServlet(holder);
+        }
+    }
+
+    public static class ForcedWebAppContext extends WebAppContext
+    {
+        {
```

Comment:
This doesnt need to be a class - you can add a bean to a standard WebAppContext.